### ISSUE TITLE
[BUG] AppInfoExt::executable panics 

### ISSUE THREAD

**Issue by jmaargh** [gtk-rs/gtk-rs-core#1073]
**Bug description**

On my system (Fedora 37, gio 0.17.4), the following demonstrates a panic from `AppInfoExt::executable`:

```rust
use gio::{traits::AppInfoExt, AppInfo};

fn main() {
    let app_info_all = AppInfo::all();

    for info in &app_info_all {
        let result = std::panic::catch_unwind(|| info.executable());
        if result.is_err() {
            println!("{}\t{}", info.id().unwrap_or("NO_ID".into()), info.name());
        }
    }
}
```

This also identifies the offending app as `qemu.desktop     QEMU`.

The backtrace points to [this debug assert](https://github.com/gtk-rs/gtk-rs-core/blob/28151a64b611e33b3eb272570501c0f390ec7bc5/glib/src/translate.rs#L1764) as a result of failing to null-check [here](https://github.com/gtk-rs/gtk-rs-core/blob/28151a64b611e33b3eb272570501c0f390ec7bc5/gio/src/auto/app_info.rs#L506).

Clearly (although the docs don't make it obvious), upstream `gio` does not guarantee a non-null response. It seems likely that other strings returned from this trait might have similar corner cases, so to guarantee soundness I would recommend null-checking them all. The choice is then, do we make the breaking change of the return types to `Option<_>`, or hotfix to some sentinel value (the empty string, an empty path, etc.)?

**Comment by bilelmoussaoui** [comment]
This should be fixed upstream by making the function annotated as nullable. Worth checking other functions as well as you mentioned.

**Comment by jmaargh** [comment]
@bilelmoussaoui I agree, but in the meantime, the gio crate is unsound. It's only a `debug_assert` that caused the panic, in a non-debug build you will have `PathBuf`s containing nulls. Since we already know that upstream can't patch this behaviour (if this data is missing, this data is missing -- and c-strings don't distinguish between the empty string and null), the rust implementation will need to change anyway.

**Comment by sdroege** [comment]
OOC, why do you consider it a worse bug than e.g. the C library handing us a dangling pointer there sometimes, or any kind of memory safety bug inside the C library. Because we can reliably detect it and work around it? It's much less bad than those cases because sooner or later it would most likely be caught by an assertion or a segfault because of a dereference, while those other cases could cause much worse outcomes.

The way forward for these kind of issues would be as follows:
  1. Create an upstream issue or (even better) MR that addresses the annotation fixes there
  2. Override the `nullable` setting on the return values in Gir.toml here and renegerate (that way we'd get `Option`s)
  3. Think about a solution for the stable branch: a) not worth it, b) change to `Option`s if it would *always* explode, c) manually implement these functions and add a proper panic, d) Same as before plus adding new functions that return an `Option`

**Comment by sdroege** [comment]
Also, did anybody (@jmaargh or @bilelmoussaoui?) start reporting this after investigating all affected functions, or plan to?

**Comment by jmaargh** [comment]
@sdroege Yes, I do consider it worse -- because the upstream issue is most likely just incorrect documentation and not actually a bug. There's nothing in the C standard that says you have to document when a `char *`-returning function might return `NULL`. However, it is invalid Rust if you construct a type with an invalid bit pattern. A soundness bug is worse than slightly incomplete documentation (especially somewhere that already has lots of incomplete documentation.

I'd be happy to PR a fix here (e.g., just returning empty string/empty path when necessary), but I wanted to check first whether that was considered desirable by the maintainers.

I did take a quick peek for upstream but frankly, for new contributors glib/etc. core libraries are not exactly easy to grok -- especially with the state of C tooling. Since a quick "gio source repository" ddg didn't even find me an obvious candidate (and the aforementioned understanding that the unsoundness flavour here is a problem for Rust where it isn't for C), I just left the report where I knew I could be helpful.

**Comment by sdroege** [comment]
> Yes, I do consider it worse -- because the upstream issue is most likely just incorrect documentation and not actually a bug.

It's not incorrect documentation but a bug on the same level as any other code bug. GObject is not plain C and the annotations are part of the code and its interface.

> However, it is invalid Rust if you construct a type with an invalid bit pattern. A soundness bug is worse than slightly incomplete documentation (especially somewhere that already has lots of incomplete documentation.

Just as it is invalid GObject code if you return `NULL` on a function that is not marked as such, or invalid C to dereference such a `NULL` later.

> I'd be happy to PR a fix here (e.g., just returning empty string/empty path when necessary), but I wanted to check first whether that was considered desirable by the maintainers.

Returning an empty string/path or some other dummy value could be a valid workaround for the stable branch too if you can find a reasonable dummy value for these cases. Otherwise a panic would be equally fine, or what I wrote above.

For the main branch these should all become `Option`s unless it's unintentional that these functions can actually return `NULL`.

The latter definitely has to happen here, together with reporting (and ideally fixing) it on the C side.

But as a very first step it's going to be required to analyze the C functions to understand if it's intentional that `NULL` is returned there and to check all the related functions if they have the same problem.

For example,. there's no point in us making the function return an `Option` if returning `NULL` is actually the bug here and the function should never do that to begin with.

> I did take a quick peek for upstream but frankly, for new contributors glib/etc. core libraries are not exactly easy to grok -- especially with the state of C tooling. Since a quick "gio source repository" ddg didn't even find me an obvious candidate

https://gitlab.gnome.org/GNOME/glib

> (and the aforementioned understanding that the unsoundness flavour here is a problem for Rust where it isn't for C), I just left the report where I knew I could be helpful.

It's also directly a bug for various other bindings that explicitly handle `NULL` in any other way than making it the programmer's problem :) It's also arguably a bug for C as the programmer would assume this function never returns `NULL` right now, and dereferencing `NULL` in C is as much UB as is dereferencing any other invalid pointers.

**Comment by jmaargh** [comment]
> But as a very first step it's going to be required to analyze the C functions to understand if it's intentional that `NULL` is returned there and to check all the related functions if they have the same problem.

Since I am not an expert in GObject-flavoured-C (TIL that comment annotations are part of the code) I am clearly not the right person to be doing this.

> Just as it is invalid GObject code if you return `NULL` on a function that is not marked as such, or invalid C to dereference such a `NULL` later.

Clearly, but FWIW the C code in the library doesn't dereference this null AFAIK, just a string-returning function that might return the empty/null string (equivalent).

I guess that you're probably a maintainer of upstream too, so I can see that you probably view these two libraries (Rust & C) as more tightly linked than me. Personally, I'd be happy to manually hotfix this crate to return empty string/empty path where necessary (since that's a valid interpretation of a null return value).

I get that this isn't ideal because of auto-generation, but I think it is necessary to hotfix in a way that doesn't change the signature (so it can be a semver patch), as the unsound code path depends on the target machine's setup and is not something that can be reliably caught by debug builds in test. There will be users right now with this unsoundness.

**Comment by sdroege** [comment]
> FWIW the C code in the library doesn't dereference this null AFAIK

Any user of the C API might, like the gio bindings here actually do :) You're not expected to check for `NULL` if the function clearly states that it doesn't return `NULL`.

> Personally, I'd be happy to manually hotfix this crate to return empty string/empty path where necessary (since that's a valid interpretation of a null return value)

It seems like a valid short-term fix for the stable branch, like I said before. At least as long as returning an empty value is not going to break callers in another way and introduce new problems. I don't know the API in question well enough to decide right now if that's a valid interpretation of `NULL` here.

The main problem is that this is the second step. First of all you need to get to the root of the problem, and that's in the C library. Randomly patching things up at a layer above without understanding the problem can easily make things worse.

FWIW, you also don't need to patch the crate. Just call that one C function directly and treat it as potentially returning `NULL`.

**Comment by jmaargh** [comment]
> FWIW, you also don't need to patch the crate. Just call that one C function directly and treat it as potentially returning `NULL`.

Yes, that's my current workaround.

> The main problem is that this is the second step. First of all you need to get to the root of the problem, and that's in the C library. Randomly patching things up at a layer above without understanding the problem can easily make things worse.

My understanding is that the proposed hotfix here is clearly an improvement. At worst, users may get some more obvious odd behaviour that leads to business-logic errors (e.g., file not found) but only in cases where they'd otherwise have UB of passing around and dereferencing `NULL` wrapped in a `NonNull` - that can only be an improvement.

This is not "randomly patching", it's responding to a function which can return `NULL`. I don't see that any more analysis is required for a hotfix. Actually marking as "nullable" (if that is indeed what's needs) would be a semver-incompatible change anyway.

Shall I make this hotfix?

**Comment by sdroege** [comment]
In the master branch that semver-incompatible change should happen, after analyzing what the correct way forward here is based on the intention of the C functions.

Once it's fixed in the master branch we can look at how to solve it for the stable branch in a somewhat backwards-compatible way.

That's exactly what I meant with randomly patching. You're suggesting to just use an empty string here without even looking at what the intended behaviour of the C function is.

We're not living by Facebook's motto here but try to do things properly.

> that can only be an improvement

Can it? Obviously garbage values that are not UB can easily cause worse outcomes when used wrongly and not being expected than a simple NULL pointer dereference that with almost 100% certainty will just kill your process.

**Comment by sdroege** [comment]
Also to be clear, the churn caused by randomly patching things now and then possibly reverting to something different later or maybe even then requiring an incompatible change in the stable branch is what I want to avoid here.

If you don't want to bother looking into the C functions then that's fine, but then this needs to wait for someone to have time to do that. Until then you have a workaround for your specific use-case.

**Comment by jmaargh** [comment]
> Can it? Obviously garbage values that are not UB can easily cause worse outcomes when used wrongly and not being expected than a simple NULL pointer dereference that with almost 100% certainty will just kill your process.

To my mind, absolutely yes. This is opinion, but even if it were just dereferencing null I don't see how a process abort is preferable to a string that happens to be empty in the vast majority of situations. More importantly, the rust compiler will make assumptions on `NonNull` values not being zero, so the potential effects are much more than dereferencing null.

> ...  randomly patching ... We're not living by Facebook's motto here but try to do things properly.

Please do not be rude to me, I am trying to help and do not appreciate these comments.

> If you don't want to bother looking into the C functions then that's fine

Again, a little rude. As I noted above, I would not consider myself qualified to do that properly since GObject-flavoured-C has special rules that I am not familiar with.

**Comment by sdroege** [comment]
> More importantly, the rust compiler will make assumptions on NonNull values not being zero, so the potential effects are much more than dereferencing null.

As does any C compiler, for that matter. A pointer that is dereferenced anywhere can't be `NULL`.

----

Sorry if what I said felt rude to you, but I don't see how patching around this without going to the root of the problem is any useful. If you don't want / can't look at the C functions then that's fine, but then this just has to wait until someone has the time to do so.

If we would just randomly patch things here we would have a lot of churn with fixes getting reverted all the time because they were actually not completely correct. You seem to have a different approach and that's fine.

I don't really have time to look into that in detail right now and this discussion already took more time than necessary so let's keep it at that.

**Comment by sdroege** [comment]
Also because having wasted so much time on this, I took a look at this specific function now. It can't return `NULL` on macOS, will return `NULL` on other UNIXes if the .desktop file has no `Exec=` line and can easily return `NULL` on Windows under various conditions.

As such, the function should simply be marked as `nullable` and that's it. Using an empty string as workaround in the stable branch of the bindings needs checking how it's actually used in practice but is probably a better approach than just panicking.

**Comment by sdroege** [comment]
Also it's clearly intended that this can be `NULL` from the check in `gdesktopappinfo.c:expand_application_parameters()` at the top.

**Comment by sdroege** [comment]
https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3369 , fix in the main branch will happen next time we sync the .gir files after that is merged. And then someone can look into a workaround for the stable branch.
